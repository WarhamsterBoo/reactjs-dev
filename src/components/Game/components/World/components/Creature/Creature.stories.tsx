import React from "react";
import { Creature } from "./Creature";
import { withKnobs, boolean } from "@storybook/addon-knobs";

export default { title: "Creature component", decorators: [withKnobs] };

export const CreatureStory: React.FC<{}> = () => {
  const isAlive = boolean("Creature is alive", false);
  return <Creature IsAlive={isAlive} />;
};
