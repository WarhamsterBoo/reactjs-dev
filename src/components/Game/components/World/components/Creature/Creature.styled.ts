import { css } from "@emotion/core";

export const BaseCreature = css`
  width: 20px;
  height: 20px;
  border: none;
  margin: 1px;
  display: inline-block;
  line-height: 25px;
  text-align: center;
  vertical-align: bottom;
`;

export const AliveCreature = css`
  background: #41ff00;
  border-radius: 5px;
`;

export const DeadCreature = css`
  background: #454a43;
`;
