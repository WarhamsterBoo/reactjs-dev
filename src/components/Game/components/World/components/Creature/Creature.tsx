import { SerializedStyles } from "@emotion/core";
import styled from "@emotion/styled";
import React, { useEffect, useState } from "react";
import { AliveCreature, BaseCreature, DeadCreature } from "./Creature.styled";

export interface CreatureProps {
  x: number;
  y: number;
  IsAlive: boolean;
  onClick: (x: number, y: number) => void;
}

export const Creature: React.FC<CreatureProps> = ({
  x,
  y,
  IsAlive,
  onClick,
}) => {
  const [creatureStyle, setCreatureStyle] = useState<SerializedStyles>(
    IsAlive ? AliveCreature : DeadCreature
  );

  useEffect(() => {
    setCreatureStyle(IsAlive ? AliveCreature : DeadCreature);
  }, [IsAlive]);

  const StyledCreature = styled.button`
    ${BaseCreature};
    ${creatureStyle};
  `;

  return <StyledCreature onClick={() => onClick(x, y)} />;
};
