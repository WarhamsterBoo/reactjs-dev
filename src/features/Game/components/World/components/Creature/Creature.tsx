import React from "react";
import { StyledCreature } from "./StyledCreature";

export interface CreatureProps {
  x: number;
  y: number;
  isAlive: boolean;
  onClick: (x: number, y: number) => void;
  transitionMs?: number;
}

export const Creature: React.FC<CreatureProps> = ({
  x,
  y,
  isAlive,
  onClick,
  transitionMs = 500,
}) => {
  return (
    <StyledCreature
      isAlive={isAlive}
      transitionMs={transitionMs}
      onClick={() => onClick(x, y)}
    />
  );
};
