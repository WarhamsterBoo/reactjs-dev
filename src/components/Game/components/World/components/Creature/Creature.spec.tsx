import { mount } from "enzyme";
import React from "react";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import { Creature, CreatureProps } from "./Creature";
import { matchers } from "jest-emotion";

expect.extend(matchers);

describe("Creature", () => {
  const defaultProps: CreatureProps = {
    x: 0,
    y: 0,
    IsAlive: false,
    onClick: jest.fn(),
    transitionMs: 500,
  };

  it("should render dead", () => {
    const sut = <Creature {...defaultProps} />;

    expect(renderer.create(sut).toJSON()).toMatchSnapshot();
  });

  it("should render alive", () => {
    const sut = <Creature {...defaultProps} IsAlive={true} />;

    expect(renderer.create(sut).toJSON()).toMatchSnapshot();
  });

  it("should start to brighten after birth", () => {
    jest.useFakeTimers();
    const sut = mount(<Creature {...defaultProps} />);
    sut.setProps({ IsAlive: true });
    sut.update();

    jest.advanceTimersByTime(100);

    sut.update();
    expect(sut).toHaveStyleRule("background", "#208000");
  });

  it("should become alive after birth", () => {
    jest.useFakeTimers();
    const sut = mount(<Creature {...defaultProps} />);
    sut.setProps({ IsAlive: true, transitionMs: 500 });
    sut.update();

    act(() => {
      jest.advanceTimersByTime(510);
    });

    sut.update();
    expect(sut).toHaveStyleRule("background", "#41ff00");
  });

  it("should start to fade after death", () => {
    jest.useFakeTimers();
    const sut = mount(<Creature {...defaultProps} IsAlive={true} />);
    sut.setProps({ IsAlive: false });
    sut.update();

    jest.advanceTimersByTime(100);

    sut.update();
    expect(sut).toHaveStyleRule("background", "#9fff80");
  });

  it("should become dead", () => {
    jest.useFakeTimers();
    const sut = mount(<Creature {...defaultProps} />);
    sut.setProps({ IsAlive: true, transitionMs: 500 });
    sut.update();

    act(() => {
      jest.advanceTimersByTime(510);
    });

    sut.update();
    expect(sut).toHaveStyleRule("background", "#41ff00");
  });

  it("should call onClick callback with it's coordinates", () => {
    const fakeOnClick = jest.fn();
    const sut = mount(
      <Creature x={1} y={2} IsAlive={true} onClick={fakeOnClick} />
    );

    sut.simulate("click");

    expect(fakeOnClick).toBeCalledTimes(1);
    expect(fakeOnClick).toBeCalledWith(1, 2);
  });
});
