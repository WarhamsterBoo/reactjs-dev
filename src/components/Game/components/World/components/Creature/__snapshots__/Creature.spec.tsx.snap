// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creature should render alive 1`] = `
.emotion-0 {
  width: 20px;
  height: 20px;
  border: none;
  margin: 1px;
  display: inline-block;
  line-height: 25px;
  text-align: center;
  vertical-align: bottom;
  background: #41ff00;
  border-radius: 5px;
}

<button
  className="emotion-0"
  onClick={[MockFunction]}
/>
`;

exports[`Creature should render dead 1`] = `
.emotion-0 {
  width: 20px;
  height: 20px;
  border: none;
  margin: 1px;
  display: inline-block;
  line-height: 25px;
  text-align: center;
  vertical-align: bottom;
  background: #454a43;
}

<button
  className="emotion-0"
  onClick={[MockFunction]}
/>
`;
